### PR TITLE
Metal compute shader

### PR DESCRIPTION
**Pull request by hana-alice**

### CODE REVIEW

**Comment by hana-alice** [comment]
@minggo pls have a review on this.

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 1]
```diff
@@ -178,23 +178,31 @@ of this software and associated engine source code (the "Software"), a limited,
...
+        _renderEncoder.setFrontFacingWinding(_gpuPipelineState->winding);
+        _renderEncoder.setDepthClipMode(_gpuPipelineState->depthClipMode);
+        _renderEncoder.setTriangleFillMode(_gpuPipelineState->fillMode);
+        _renderEncoder.setRenderPipelineState(_gpuPipelineState->mtlRenderPipelineState);
+
+        if (_gpuPipelineState->mtlDepthStencilState) {
+            _renderEncoder.setStencilFrontBackReferenceValue(_gpuPipelineState->stencilRefFront, _gpuPipelineState->stencilRefBack);
+            _renderEncoder.setDepthStencilState(_gpuPipelineState->mtlDepthStencilState);
+        }
+    } else if (pso->getBindPoint() == PipelineBindPoint::COMPUTE) {
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:200]: Can only invoke `pso->getBindPoint()` once.

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 2]
```diff
@@ -507,24 +523,67 @@ of this software and associated engine source code (the "Software"), a limited,
...
+        
+        if(sampler.stages & ShaderStageFlagBit::COMPUTE) {
+            _computeEncoder.setTexture(gpuDescriptor.texture->getMTLTexture(), sampler.textureBinding);
         }
     }
 }
 
 void CCMTLCommandBuffer::blitTexture(Texture *srcTexture, Texture *dstTexture, const TextureBlit *regions, uint count, Filter filter) {
+    id<MTLBlitCommandEncoder> encoder = [_mtlCommandBuffer blitCommandEncoder];
+    id<MTLTexture> src = ((CCMTLTexture*)srcTexture)->getMTLTexture();
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:543]: Please don't use c style conversion.

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 3]
```diff
@@ -507,24 +523,67 @@ of this software and associated engine source code (the "Software"), a limited,
...
+        dst = [currentDrawable texture];
+    } else {
+        dst = ((CCMTLTexture*)dstTexture)->getMTLTexture();
+    }
+    
+    if([src pixelFormat] != device->preferredPixelFormat()) {
+        src = [src newTextureViewWithPixelFormat:(MTLPixelFormat)device->preferredPixelFormat()];
+    }
+    //
+    for (int i = 0; i < count; ++i) {
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:557]: count is uint, will have warning

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 4]
```diff
@@ -507,24 +523,67 @@ of this software and associated engine source code (the "Software"), a limited,
...
+        CAMetalLayer* currentDrawable = static_cast<CAMetalLayer*>(device->getCurrentDrawable());
+        dst = [currentDrawable texture];
+    } else {
+        dst = ((CCMTLTexture*)dstTexture)->getMTLTexture();
+    }
+    
+    if([src pixelFormat] != device->preferredPixelFormat()) {
+        src = [src newTextureViewWithPixelFormat:(MTLPixelFormat)device->preferredPixelFormat()];
+    }
+    //
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:556]: should delete it?

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 5]
```diff
@@ -507,24 +523,67 @@ of this software and associated engine source code (the "Software"), a limited,
...
+    id<MTLBlitCommandEncoder> encoder = [_mtlCommandBuffer blitCommandEncoder];
+    id<MTLTexture> src = ((CCMTLTexture*)srcTexture)->getMTLTexture();
+    id<MTLTexture> dst = nil;
+    CCMTLDevice* device = static_cast<CCMTLDevice*>(_device);
+    if(!dstTexture) {
+        CAMetalLayer* currentDrawable = static_cast<CAMetalLayer*>(device->getCurrentDrawable());
+        dst = [currentDrawable texture];
+    } else {
+        dst = ((CCMTLTexture*)dstTexture)->getMTLTexture();
+    }
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:551]: Can modify current drawable? When need to do like this?

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 6]
```diff
@@ -507,24 +523,67 @@ of this software and associated engine source code (the "Software"), a limited,
...
+        }
+        
+        if(sampler.stages & ShaderStageFlagBit::COMPUTE) {
+            _computeEncoder.setTexture(gpuDescriptor.texture->getMTLTexture(), sampler.textureBinding);
         }
     }
 }
 
 void CCMTLCommandBuffer::blitTexture(Texture *srcTexture, Texture *dstTexture, const TextureBlit *regions, uint count, Filter filter) {
+    id<MTLBlitCommandEncoder> encoder = [_mtlCommandBuffer blitCommandEncoder];
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:542]: It is better to check `srcTexture` and `regions` too.

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm (hunk 5 above)
Comment (hana-alice) [reply cocos/renderer/gfx-metal/MTLCommandBuffer.mm:551]: gfxtestcase:
```
    if (_device->hasFeature(gfx::Feature::COMPUTE_SHADER)) {
        commandBuffer->pipelineBarrier(_globalBarriers[1], _textureBarriers.data(), _textures.data(), 1);

        commandBuffer->bindPipelineState(_compPipelineState);
        commandBuffer->bindDescriptorSet(0, _compDescriptorSet);
        commandBuffer->dispatch(dispatchInfo);

        commandBuffer->bindPipelineState(_compBGPipelineState);
        commandBuffer->bindDescriptorSet(0, _compBGDescriptorSet);
        commandBuffer->dispatch(bgDispatchInfo);

#if USE_BLIT_TEXTURE
        commandBuffer->pipelineBarrier(_globalBarriers[2], &_textureBarriers[1], _textures.data(), 2);
        commandBuffer->blitTexture(_textures[0], nullptr, &blit, 1u, gfx::Filter::POINT);
#endif
    }
```

when second parameters of blitTexture == null, command buffer blit the texture to swapchain/drawable/framebuffer

File: cocos/renderer/gfx-metal/MTLComputeCommandEncoder.h [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        [_mtlEncoder setBuffer:buffer offset:offset atIndex:index];
+    }
+
+    CC_INLINE void setTexture(const id<MTLTexture> texture, uint index) {
+        [_mtlEncoder setTexture:texture atIndex:index];
+        _resourceSize = {texture.width, texture.height, texture.depth};
+    }
+    
+    CC_INLINE void dispatch(MTLSize groupsPerGrid) {
+        // nothing to do with thread count in GLSL shader
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLComputeCommandEncoder.h:73]: GLSL shader?

File: cocos/renderer/gfx-metal/MTLPipelineState.mm [hunk 8]
```diff
@@ -85,24 +86,50 @@ of this software and associated engine source code (the "Software"), a limited,
         CC_LOG_ERROR("CCMTLPipelineState: CC_NEW CCMTLGPUPipelineState failed.");
         return false;
     }
+    
+    if(_bindPoint == PipelineBindPoint::GRAPHICS) {
+        if (!createMTLRenderPipelineState()) {
+            return false;
+        }
+        // Application can run with wrong depth/stencil state.
+        if (!createMTLDepthStencilState()) return false;
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLPipelineState.mm:95]: Please use the same code format for if statement. Can use
```
    if () xxx
   
    if ()
        xxx


    if () {
        xxx
    }
```
But please use the same format in a file.

File: cocos/renderer/gfx-metal/MTLShader.mm [hunk 9]
```diff
@@ -90,26 +106,36 @@ of this software and associated engine source code (the "Software"), a limited,
...
         }
+    } else if (isComputeShader) {
+        _computeMTLFunction = [library newFunctionWithName:@"main0"];
+        if (!_computeMTLFunction) {
+            [library release];
+            CC_LOG_ERROR("Can not create compute function: main0");
+            return false;
+        }
+    } else {
+        CC_LOG_ERROR("Shader type not supported yet!");
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLShader.mm:138]: Just return false here?

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 10]
```diff
@@ -507,24 +524,69 @@ of this software and associated engine source code (the "Software"), a limited,
...
 
 void CCMTLCommandBuffer::blitTexture(Texture *srcTexture, Texture *dstTexture, const TextureBlit *regions, uint count, Filter filter) {
+    if(srcTexture && regions) {
+        id<MTLBlitCommandEncoder> encoder = [_mtlCommandBuffer blitCommandEncoder];
+        id<MTLTexture> src = static_cast<CCMTLTexture*>(srcTexture)->getMTLTexture();
+        id<MTLTexture> dst = nil;
+        CCMTLDevice* device = static_cast<CCMTLDevice*>(_device);
+        if(!dstTexture) {
+            id<CAMetalDrawable> currentDrawable = (id<CAMetalDrawable>)device->getCurrentDrawable();
+            dst = [currentDrawable texture];
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:550]: Is it designed like this? Change the content of onscreen texture with blit?

File: cocos/renderer/gfx-metal/MTLComputeCommandEncoder.h [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    CCMTLComputeCommandEncoder &operator=(CCMTLComputeCommandEncoder &&) = delete;
+
+    void initialize(id<MTLCommandBuffer> commandBuffer) {
+        _mtlEncoder  = [[commandBuffer computeCommandEncoder] retain];
+        _initialized = true;
+    }
+
+    bool isInitialized() {
+        return _initialized;
+    }
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLComputeCommandEncoder.h:54]: It is better to add const.

File: cocos/renderer/gfx-metal/MTLCommandEncoder.h [hunk 12]
```diff
@@ -0,0 +1,38 @@
+/****************************************************************************
+ Copyright (c) 2019-2021 Xiamen Yaji Software Co., Ltd.
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandEncoder.h:2]: 2019-2021 -> 2021, as this file is a new file.

File: cocos/renderer/gfx-metal/MTLComputeCommandEncoder.h (hunk 7 above)
Comment (hana-alice) [reply cocos/renderer/gfx-metal/MTLComputeCommandEncoder.h:73]: GLSL shader request to specify the compute thread size, no such limit in Metal and have to set compute thread size explicity, so clarified here

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm (hunk 10 above)
Comment (hana-alice) [reply cocos/renderer/gfx-metal/MTLCommandBuffer.mm:550]: before present() is called, it's still an offscreen texture, we haven't hold any pointer for presented texture. It's just a texture belongs to drawable, and it's legal to blit texture to drawable: https://developer.apple.com/forums/thread/63589

Comment (minggo) [reply cocos/renderer/gfx-metal/MTLCommandBuffer.mm:550]: Good to know, thanks.

**Review by minggo — APPROVED** [review]

**Pull request merged**